File: backend/src/models/Transaction.js
const mongoose = require('mongoose');
const crypto = require('crypto');

const TransactionSchema = new mongoose.Schema(
  {
    transactionId: {
      type: String,
      default: () => crypto.randomUUID(),
      unique: true,
      index: true
    },
    userId: {
      type: String,
      ref: 'User',
      required: [true, 'User ID is required'],
      index: true
    },
    amount: {
      type: Number,
      required: [true, 'Amount is required']
    },
    category: {
      type: String,
      required: [true, 'Category is required'],
      enum: {
        values: [
          'Food', 'Transport', 'Shopping', 'Bills', 'Entertainment',
          'Healthcare', 'Education', 'Rent', 'Utilities', 'Salary',
          'Investment', 'Others'
        ],
        message: '{VALUE} is not a supported category'
      }
    },
    type: {
      type: String,
      required: [true, 'Transaction type is required'],
      enum: {
        values: ['income', 'expense'],
        message: '{VALUE} is not a valid transaction type'
      }
    },
    date: {
      type: Date,
      default: Date.now,
      index: true
    },
    note: {
      type: String,
      trim: true,
      default: ""
    }
  },
  {
    timestamps: true,
    toJSON: {
      transform: function (doc, ret) {
        delete ret._id;
        delete ret.__v;
        return ret;
      }
    }
  }
);

module.exports = mongoose.model('Transaction', TransactionSchema);

File: flutter_app/lib/models/transaction.dart
class TransactionModel {
  final String transactionId;
  final String userId;
  final double amount;
  final String category;
  final String type; // 'income' or 'expense'
  final DateTime date;
  final String note;
  final DateTime? createdAt;

  TransactionModel({
    required this.transactionId,
    required this.userId,
    required this.amount,
    required this.category,
    required this.type,
    required this.date,
    required this.note,
    this.createdAt,
  });

  factory TransactionModel.fromJson(Map<String, dynamic> json) {
    return TransactionModel(
      transactionId: json['transactionId'] ?? '',
      userId: json['userId'] ?? '',
      amount: (json['amount'] ?? 0).toDouble(),
      category: json['category'] ?? '',
      type: json['type'] ?? 'expense',
      date: json['date'] != null ? DateTime.parse(json['date']) : DateTime.now(),
      note: json['note'] ?? '',
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'transactionId': transactionId,
      'userId': userId,
      'amount': amount,
      'category': category,
      'type': type,
      'date': date.toIso8601String(),
      'note': note,
      'createdAt': createdAt?.toIso8601String(),
    };
  }
}
